File: week_8_and_9/3_Ruby/fibonacci_sequence/my_solution.rb
# U3.W8-9: 


# I worked on this challenge [by myself, with: ].

# 2. Pseudocode
=begin
 
def is_fibonacci?(num)
   array = [0..num]
    new_array =[]
     array.each_with_index do |i, index|
        if array[index] = array[index-1] + array[index-2]
            then new_array << array[index]
        end
     end
    if array.include?(num)
        true
    else
        false
    end
end

 
 
 
end
 
=end



# 3. Initial Solution

=begin
def is_fibonacci?(num)
    array = (0..num).to_a
    counter = 0
    fib = []
      while counter < array.length
          counter += 1
            if array[counter] == array[counter-1] + array[counter-2]
              fib << array[counter]
            end
          counter += 1
      end
    puts fib
    fib.include?(num)
end
=end






=begin
def is_fibonacci?(num)
    array = (0..num).to_a
    fib = []
    array.each_with_index do |i, index|
        if array[index] == array[index-1] + array[index-2]
            fib << array[index]
        end
    end
    puts fib
    fib.include?(num)
end
=end

def is_fibonacci?(num)
  fib = [0,1]

  while fib[-1] < num
    fib << fib[-1] + fib[-2]
  end


  fib[-1] == num ? true: false

end


#puts is_fibonacci?(5)
#puts is_fibonacci?(4)
# 4. Refactored Solution






# 1. DRIVER TESTS GO BELOW THIS LINE

 def assert
     raise "Assertion failed!" unless yield
 end

 def random_fibonacci
     [0,1,1,2,3,5,8,13,21,34,55,89,144,233,377,610,987,1597,2584,4181,6765,10946].sample
 end

 assert {is_fibonacci?(random_fibonacci) == true}

 assert {is_fibonacci?(8670007398507948658051921) == true}
 assert {is_fibonacci?(random_fibonacci+100) == false}
 assert {is_fibonacci?(927372692193078999171) == false}






# 5. Reflection 

File: week_8_and_9/2_JavaScript/grocery_list/my_solution.js
// U3.W8-9: 


// I worked on this challenge [by myself, with: ].

// 2. Pseudocode
//ATTEMPT 1- Psedocode
// var list ={}

// console.log(list)

// list.new_list = {}

// list.new_item[item] = name;


// list.addItem = function(item, category){
//     list.new_list[item] = category;
// }

// list.deleteItem = function(item, category){
// 	delete list.new_list[item];
// }

// list.viewList = function(){
// 	console.log(list.new_list);
// }

// list.deleteList = function(){
//   list.new_list = {};
// }

//ATTEMPT 2 Pseudocode
// var list = {
//   list.banana = {};
//   list.apple = {};
//   list.milk = {};
// };

// function Item(name, category, checked) {
//     this.name = name;
//     this.category = category;
//     this.checked = false;
    
// }

// var banana = new Item("banana", "fruit", false)
// var milk = new Item("milk", "dairy", false)
// var chocolate = new Item("chocolate", "sweets", false)
// var steak = new Item("steak", "meat", false)

// console.log(banana)
// console.log(milk)

//var list = [banana, milk, chocolate, steak];

// for (i=0; i < list.length; i++){
//   console.log(list[i]);
// }

// banana.checked = true
// console.log(banana.checked)





// 3. Initial Solution (FIRST ATTEMPT)

// var list ={}

// console.log(list)

// list.new_list = {}

// list.new_item[item] = name;


// list.addItem = function(item, category){
//     list.new_list[item] = category;
// }

// list.deleteItem = function(item, category){
// 	delete list.new_list[item];
// }

// list.viewList = function(){
// 	console.log(list.new_list);
// }

// list.deleteList = function(){
//   list.new_list = {};
// }


// }

// list.addItem("Cherries","Fruit")
// // console.log(list.new_list)
// // console.log(list)
// list.viewList()

// // console.log(list.category("Fruit"))
// list.deleteList()
// list.viewList()
// // list.deleteItem("Cherries", "Fruit")
// // console.log(list.new_list)



// 3. Initial Solution (Second ATTEMPT)

var groceryList = [];

var Item = function(name, category, checked) {
    this.name = name;
    this.category = category;
    this.checked = false;
    
}

var addItem = function(name) {
  for (var i=0; i < arguments.length; i++) {
    groceryList.push(arguments[i]);
  }
}

var removeItem = function(name) {
  var index = groceryList.indexOf(name);
    if (index >= 0) {
	  groceryList.splice(index, 1);
	}
}
// display the items in a specific category
var viewList = function(category) {
  for (i = 0; i < groceryList.length; i++) {
    if (category === groceryList[i].category) {
	  console.log(groceryList[i].category + ' : ' + groceryList[i].name);
     }
   }
}

var checked = function(name) {
  for (i=0; i<groceryList.length;i++) {
	if (name === groceryList[i].name) {
		console.log(groceryList[i].checked === true);
	}
  }
}


var apple = new Item("apple","fruit",false);
var steak = new Item("steak","meat",false);
addItem(apple);
addItem(steak);
//console.log(groceryList);
//viewList("fruit");
checked(apple);
console.log(groceryList);


// 4. Refactored Solution






// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE






// 5. Reflection 